doc/md/canister-maintenance/upgrades.md

```diff
@@ -0,0 +1,154 @@
+
+The only difference between stable types and shared types is the former’s support for mutation. Like shared types, stable types are restricted to first-order data, excluding local functions and structures built from local functions, such as objects. This exclusion of functions is required because the meaning of a function value, consisting of both data and code, cannot easily be preserved across an upgrade. The meaning of plain data, mutable or not, can be.
+
+:::note
+
+In general, object types are not stable because they can contain local functions. However, a plain record of stable data is a special case of object types that are stable. Moreover, references to actors and shared functions are also stable, allowing you to preserve their values across upgrades. For example, you can preserve the state record of a set of actors or shared function callbacks subscribing to a service.
+
+:::
+
+## Stable type signatures
+
+The collection of stable variable declarations in an actor can be summarized in a stable signature.
+
+The textual representation of an actor’s stable signature resembles the internals of a Motoko actor type:
+
+``` motoko no-repl
+actor {
+  stable x : Nat;
+  stable var y : Int;
+  stable z : [var Nat];
+};
+```
+
+It specifies the names, types and mutability of the actor’s stable fields, possibly preceded by relevant Motoko type declarations.
+
+:::tip
+
+You can emit the stable signature of the main actor or actor class to a `.most` file using `moc` compiler option `--stable-types`. You should never need to author your own `.most` file.
+
+:::
+
+A stable signature `<stab-sig1>` is stable-compatible with signature `<stab-sig2>`, if:
+
+-   Every immutable field `stable <id> : T` in `<stab-sig1>` has a matching field `stable <id> : U` in `<stab-sig2>` with `T <: U`.
+
+-   Every mutable field `stable var <id> : T` in `<stab-sig1>` has a matching field `stable var <id> : U` in `<stab-sig2>` with `T <: U`.
+
+Note that `<stab-sig2>` may contain additional fields. Typically, `<stab-sig1>` is the signature of an older version while `<stab-sig2>` is the signature of a newer version.
+
+The subtyping condition on stable fields ensures that the final value of some field can be consumed as the initial value of that field in the upgraded code.
+
+:::tip
+
+You can check the stable-compatibility of two `.most` files containing stable signatures, using `moc` compiler option `--stable-compatible file1.most file2.most`.
+
+:::
+
+:::note
+
+The stable-compatible relation is quite conservative. In the future, it may be relaxed to accommodate a change in field mutability and/or abandoning fields from `<stab-sig1>` but with a warning.
+
+:::
+
+## Preupgrade and postupgrade system methods
```

Comment:
I would maybe move this back just after the Section on ##Typing since its really about how to persist data that isn't inherently stable. Then the rest of the section doc is about upgrade safety and how to check it.